File: plugins/plugin-discord/typescript/actions/sendDM.ts
import {
  type Action,
  type ActionExample,
  type ActionResult,
  type Content,
  composePromptFromState,
  type HandlerCallback,
  type HandlerOptions,
  type IAgentRuntime,
  type Memory,
  ModelType,
  parseJSONObjectFromText,
  type State,
} from "@elizaos/core";
import type { User } from "discord.js";
import { DISCORD_SERVICE_NAME } from "../constants";
// Import generated prompts
import { sendDmTemplate } from "../generated/prompts/typescript/prompts.js";
import { requireActionSpec } from "../generated/specs/spec-helpers";
import type { DiscordService } from "../service";

/**
 * Get DM information from the user's request
 * @param {IAgentRuntime} runtime - The runtime object to interact with the agent.
 * @param {Memory} _message - The memory object containing the input message.
 * @param {State} state - The state of the conversation.
 * @returns {Promise<{recipientIdentifier: string, messageContent: string} | null>} DM info or null if not parseable.
 */
const getDMInfo = async (
  runtime: IAgentRuntime,
  _message: Memory,
  state: State
): Promise<{ recipientIdentifier: string; messageContent: string } | null> => {
  const prompt = composePromptFromState({
    state,
    template: sendDmTemplate,
  });

  for (let i = 0; i < 3; i++) {
    const response = await runtime.useModel(ModelType.TEXT_SMALL, {
      prompt,
    });

    const parsedResponse = parseJSONObjectFromText(response) as {
      recipientIdentifier: string;
      messageContent: string;
    } | null;

    if (parsedResponse?.recipientIdentifier && parsedResponse.messageContent) {
      return parsedResponse;
    }
  }
  return null;
};

/**
 * Find a Discord user by various identifiers
 * @param {DiscordService} discordService - The Discord service instance
 * @param {string} identifier - The user identifier (username, ID, or mention)
 * @param {string} currentServerId - The current server ID to search in
 * @returns {Promise<User | null>} The found user or null
 */
const findUser = async (
  discordService: DiscordService,
  identifier: string,
  currentServerId?: string
): Promise<User | null> => {
  if (!discordService.client) {
    return null;
  }

  // Remove mention formatting if present
  const cleanId = identifier.replace(/[<@!>]/g, "");

  try {
    // Try to fetch by ID first
    if (/^\d+$/.test(cleanId)) {
      try {
        return await discordService.client.users.fetch(cleanId);
      } catch (_e) {
        // ID not found, continue to username search
      }
    }

    // Search in the current server if available
    if (currentServerId) {
      const guild = await discordService.client.guilds.fetch(currentServerId);
      const members = await guild.members.fetch();

      // Search by username or display name
      const member = members.find(
        (m) =>
          m.user.username.toLowerCase() === identifier.toLowerCase() ||
          m.displayName.toLowerCase() === identifier.toLowerCase() ||
          m.user.tag.toLowerCase() === identifier.toLowerCase()
      );

      if (member) {
        return member.user;
      }
    }

    // Search in all guilds the bot is in
    const guilds = Array.from(discordService.client.guilds.cache.values());
    for (const guild of guilds) {
      try {
        const members = await guild.members.fetch();
        const member = members.find(
          (m) =>
            m.user.username.toLowerCase() === identifier.toLowerCase() ||
            m.displayName.toLowerCase() === identifier.toLowerCase() ||
            m.user.tag.toLowerCase() === identifier.toLowerCase()
        );

        if (member) {
          return member.user;
        }
      } catch (_e) {
        // Continue searching in other guilds
      }
    }

    return null;
  } catch (_error) {
    // Standalone function - error handled by caller
    return null;
  }
};

const spec = requireActionSpec("SEND_DM");

export const sendDM: Action = {
  name: spec.name,
  similes: spec.similes ? [...spec.similes] : [],
  description: spec.description,
  validate: async (_runtime: IAgentRuntime, message: Memory, _state?: State): Promise<boolean> => {
    return message.content.source === "discord";
  },
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback
  ): Promise<ActionResult | undefined> => {
    const discordService = runtime.getService(DISCORD_SERVICE_NAME) as DiscordService;

    if (!discordService || !discordService.client) {
      runtime.logger.error(
        { src: "plugin:discord:action:send-dm", agentId: runtime.agentId },
        "Discord service not found or not initialized"
      );
      return { success: false, error: "Discord service is not available" };
    }

    if (!state) {
      if (callback) {
        await callback?.({
          text: "State is not available.",
          source: "discord",
        });
      }
      return { success: false, error: "State is not available" };
    }

    const dmInfo = await getDMInfo(runtime, message, state);
    if (!dmInfo) {
      runtime.logger.warn(
        { src: "plugin:discord:action:send-dm", agentId: runtime.agentId },
        "Could not parse DM information from message"
      );
      if (callback) {
        await callback?.({
          text: "I couldn't understand who you want me to message or what to send. Please specify the recipient and the message content.",
          source: "discord",
        });
      }
      return { success: false, error: "Could not parse DM information" };
    }

    try {
      const room = state.data?.room || (await runtime.getRoom(message.roomId));
      const currentServerId = room?.messageServerId;

      // Find the user
      const targetUser = await findUser(
        discordService,
        dmInfo.recipientIdentifier,
        currentServerId
      );

      if (!targetUser) {
        if (callback) {
          await callback?.({
            text: `I couldn't find a user with the identifier "${dmInfo.recipientIdentifier}". Please make sure the username or ID is correct.`,
            source: "discord",
          });
        }
        return { success: false, error: `User not found: ${dmInfo.recipientIdentifier}` };
      }

      // Check if we can send DMs to this user
      if (targetUser.bot) {
        if (callback) {
          await callback?.({
            text: "I cannot send direct messages to other bots.",
            source: "discord",
          });
        }
        return { success: false, error: "Cannot send DMs to bots" };
      }

      // Create or get DM channel
      const dmChannel = await targetUser.createDM();

      // Send the message
      await dmChannel.send(dmInfo.messageContent);

      const response: Content = {
        text: `I've sent your message to ${targetUser.username}: "${dmInfo.messageContent}"`,
        actions: ["SEND_DM_RESPONSE"],
        source: message.content.source,
      };

      if (callback) {
        await callback?.(response);
      }
      return { success: true, text: response.text };
    } catch (error) {
      runtime.logger.error(
        {
          src: "plugin:discord:action:send-dm",
          agentId: runtime.agentId,
          error: error instanceof Error ? error.message : String(error),
        },
        "Error sending DM"
      );

      // Handle specific Discord API errors
      if (error instanceof Error) {
        if (error.message.includes("Cannot send messages to this user")) {
          if (callback) {
            await callback?.({
              text: "I couldn't send a message to that user. They may have DMs disabled or we don't share a server.",
              source: "discord",
            });
          }
        } else {
          if (callback) {
            await callback?.({
              text: "I encountered an error while trying to send the direct message. Please make sure I have the necessary permissions.",
              source: "discord",
            });
          }
        }
      }
      return { success: false, error: error instanceof Error ? error.message : String(error) };
    }
  },
  examples: (spec.examples ?? []) as ActionExample[][],
};

export default sendDM;

File: plugins/plugin-nostr/python/elizaos_plugin_nostr/actions/send_dm.py
"""
Send DM action for Nostr plugin.
"""

import json
import logging
import re

from ..types import (
    NOSTR_SERVICE_NAME,
    NostrDmSendOptions,
    is_valid_pubkey,
    normalize_pubkey,
    split_message_for_nostr,
)

logger = logging.getLogger(__name__)

SEND_DM_TEMPLATE = """# Task: Extract Nostr DM parameters
Based on the conversation, determine what message to send and to whom.

Recent conversation:
{recent_messages}

Extract the following:
- text: The message content to send
- toPubkey: The target pubkey (npub or hex format, or "current" for the current conversation)

Respond with a JSON object:
```json
{{
  "text": "message content here",
  "toPubkey": "npub1... or hex pubkey or current"
}}
```"""


async def validate(runtime, message, state: dict | None = None) -> bool:
    """Validate that this action can be executed."""
    return message.content.get("source") == "nostr"


async def handler(
    runtime,
    message,
    state: dict | None = None,
    options: dict | None = None,
    callback=None,
):
    """Handle the send DM action."""
    nostr_service = runtime.get_service(NOSTR_SERVICE_NAME)

    if not nostr_service or not nostr_service.is_connected():
        if callback:
            await callback({"text": "Nostr service is not available.", "source": "nostr"})
        return {"success": False, "error": "Nostr service not available"}

    # Build prompt
    recent_messages = state.get("recentMessages", "") if state else ""
    prompt = SEND_DM_TEMPLATE.format(recent_messages=recent_messages)

    # Extract parameters using LLM
    dm_info = None
    for _ in range(3):
        response = await runtime.use_model("TEXT_SMALL", {"prompt": prompt})

        try:
            json_match = re.search(r"\{[^{}]*\}", response, re.DOTALL)
            if json_match:
                parsed = json.loads(json_match.group())
                if parsed.get("text"):
                    dm_info = {
                        "text": str(parsed["text"]),
                        "to_pubkey": str(parsed.get("toPubkey", "current")),
                    }
                    break
        except (json.JSONDecodeError, ValueError):
            continue

    if not dm_info or not dm_info.get("text"):
        if callback:
            await callback({
                "text": "I couldn't understand what message you want me to send.",
                "source": "nostr",
            })
        return {"success": False, "error": "Could not extract message parameters"}

    # Determine target pubkey
    target_pubkey = None
    if dm_info["to_pubkey"] and dm_info["to_pubkey"] != "current":
        if is_valid_pubkey(dm_info["to_pubkey"]):
            try:
                target_pubkey = normalize_pubkey(dm_info["to_pubkey"])
            except Exception:
                pass

    # Get pubkey from state if available
    if not target_pubkey:
        state_data = state.get("data", {}) if state else {}
        target_pubkey = state_data.get("senderPubkey")

    if not target_pubkey:
        if callback:
            await callback({
                "text": "I couldn't determine who to send the message to.",
                "source": "nostr",
            })
        return {"success": False, "error": "Could not determine target pubkey"}

    # Split message if too long
    chunks = split_message_for_nostr(dm_info["text"])

    # Send message(s)
    last_result = None
    for chunk in chunks:
        opts = NostrDmSendOptions(to_pubkey=target_pubkey, text=chunk)
        result = await nostr_service.send_dm(opts)

        if not result.success:
            if callback:
                await callback({
                    "text": f"Failed to send message: {result.error}",
                    "source": "nostr",
                })
            return {"success": False, "error": result.error}

        last_result = {"event_id": result.event_id, "relays": result.relays}

    if callback:
        await callback({
            "text": "Message sent successfully.",
            "source": message.content.get("source"),
        })

    return {
        "success": True,
        "data": {
            "to_pubkey": target_pubkey,
            "event_id": last_result.get("event_id") if last_result else None,
            "relays": last_result.get("relays") if last_result else [],
            "chunks_count": len(chunks),
        },
    }


send_dm_action = {
    "name": "NOSTR_SEND_DM",
    "similes": [
        "SEND_NOSTR_DM",
        "NOSTR_MESSAGE",
        "NOSTR_TEXT",
        "DM_NOSTR",
    ],
    "description": "Send an encrypted direct message via Nostr (NIP-04)",
    "validate": validate,
    "handler": handler,
    "examples": [
        [
            {"name": "{{user1}}", "content": {"text": "Send them a message saying 'Hello!'"}},
            {
                "name": "{{agent}}",
                "content": {
                    "text": "I'll send that DM via Nostr.",
                    "actions": ["NOSTR_SEND_DM"],
                },
            },
        ],
    ],
}
